kmscon: unstable-2018-09-07 -> 9.0.0

Infinidoge opened the pull request:
###### Description of changes

This updates kmscon to the latest release tag on the repository, after 3 years(!) without an update in Nixpkgs.

NOTE: This PR depends on  #171069

Notifying maintainer; @omasanori 

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Infinidoge committed:
kmscon: unstable-2018-09-07 -> 9.0.0

Infinidoge commented:
As noted in #178748, updating meson appears to have created a small variety of build errors due to the `fortify` hardening. It seems that meson is removing the `-O2` flag from the hardening, causing the compiler to reject it due to the `fortify` hardening requiring the optimization flag.

Infinidoge commented:
Was made aware that someone else made a meson PR, so retargetting this on top of that.

Infinidoge commented:
Rebased on top of the meson changes and edited the PR message accordingly. 

I didn't see anything in the Nixpkgs manual about the case of depending on another PR being merged, so my (current) assumption is that when the meson change is merged, the commits this PR has will be seen as already included and thus won't have duplicates, but I could very much be wrong. 

Please let me know if I should just rebase out the commits this depends on.

omasanori reviewed (commented):
Regarding Meson changes, I think it would be better to keep them out of this PR and wait for the separate PR to be merged.

Infinidoge commented:
Rebased out the dependent commits.

AndersonTorres commented on pkgs/os-specific/linux/kmscon/default.nix:
```diff
@@ -9,19 +9,22 @@
 stdenv.mkDerivation rec {
   pname = "kmscon";
-  version = "unstable-2018-09-07";
+  version = "v9.0.0";
```
```suggestion
  version = "9.0.0";
```

Infinidoge replied:
Ah, good point, I should have checked that. Fixed.

AndersonTorres commented on pkgs/os-specific/linux/kmscon/default.nix:
```diff
@@ -9,19 +9,22 @@
     repo = "kmscon";
-    rev = "01dd0a231e2125a40ceba5f59fd945ff29bf2cdc";
-    sha256 = "0q62kjsvy2iwy8adfiygx2bfwlh83rphgxbis95ycspqidg9py87";
+    rev = version;
```
```suggestion
    rev = v${version};
```

AndersonTorres commented on pkgs/os-specific/linux/kmscon/default.nix:
```diff
@@ -33,14 +36,21 @@ stdenv.mkDerivation rec {
     pixman
     systemd
+    mesa
+    cmake
```
Usually cmake is not recommended when using Meson+Ninja. What are you doing here?

Infinidoge replied:
I believe I added that when I was initially debugging the building of the package after updating, since it mentioned cmake being missing. Will remove and test building.

AndersonTorres replied:
This warning message is harmless.

When Meson tries to find the dependencies described in its `meson.build`, it tries to call some programs and scripts that can help it to find them.
One of them is cmake.

It is more useful for developers working with multiple softwares; for us package writers (especially Nixpkgs, that forces you to include every dependent package), it makes little difference.

AndersonTorres replied:
Reading the logs, I have noticed that the error is because our meson is still outdated.

Infinidoge replied:
Ahhh, I see now. Removing it now and force-pushing.

Infinidoge commented:
Pulled in a patch from after the most recent release, and added a compiler flag which disables a specific warning, as those warnings (treated as errors) prevented compilation. Also as noted in the previous review, removed `cmake` as an extraneous dependency. See the recently mentioned issue for the issue behind the compiler flag.

AndersonTorres commented on pkgs/os-specific/linux/kmscon/default.nix:
```diff
@@ -33,14 +36,22 @@ stdenv.mkDerivation rec {
     pkg-config
   ];
 
+  patches = [ ./0001-tests-fix-warnings.patch ]; # https://github.com/Aetf/kmscon/commit/b65f4269b03de580923ab390bde795e7956b633f
```
Put the comments before the patches. It is more readable.
```
  patches = [ 
    # https://github.com/Aetf/kmscon/commit/b65f4269b03de580923ab390bde795e7956b633f
    ./0001-tests-fix-warnings.patch 
  ];
```
----------

Way better, in this case you should use fetchpatch

```suggestion
  patches = [
    (fetchpatch {
       name = "0001-tests-fix-warnings.patch";
       # Below is a trick: add `.patch` extension after the GitHub URL, and it will show you a raw patch file
       url = "https://github.com/Aetf/kmscon/commit/b65f4269b03de580923ab390bde795e7956b633f.patch";
       sha256 = ""; # run the expression, and the error message will tell you the true, expected hash 
     };)
  ];

Infinidoge replied:
You know, I remember trying to get the `.patch` of a commit like that via the URL and it just didn't work, I guess I didn't quite use the right URL. That is definitely the better method, updating the PR now.

Infinidoge commented:
Incidentally included the auto-formatting my editor did for lines 1/2, 10/11, and 33/34, which I'm going to leave in since I personally prefer the expanded form. If I should revert that shift, please let me know.

jtojnar commented on pkgs/os-specific/linux/kmscon/default.nix:
```diff
@@ -1,46 +1,67 @@
     pixman
     systemd
+    mesa
+    ninja
```
`ninja` belongs to `nativeBuildInputs`.

AndersonTorres commented on pkgs/os-specific/linux/kmscon/default.nix:
```diff
@@ -1,46 +1,67 @@
   ];
 
+  patches = [
+    # https://github.com/Aetf/kmscon/commit/b65f4269b03de580923ab390bde795e7956b633f
```
```suggestion
```

Useless comment.

Infinidoge replied:
Oh, true. Removed.

omasanori commented:
> Incidentally included the auto-formatting my editor did for lines 1/2, 10/11, and 33/34, which I'm going to leave in since I personally prefer the expanded form. If I should revert that shift, please let me know.

It is acceptable to me.

By the way, what kind of tests have you done? Just completed a build or running the kmscon service with this? While I will do some tests this weekend anyways, I'd know what is succeeded on your machine.

Infinidoge commented:
For those who want to more easily test the changes using flakes, I've made a branch on my nixpkgs fork that combines the meson and kmscon changes: https://github.com/Infinidoge/nixpkgs/tree/combined/kmscon
(`github:Infinidoge/nixpkgs/combined/kmscon`)

Infinidoge commented:
Seems to be running just fine on my machine, which is good:
![image](https://user-images.githubusercontent.com/22727114/176986933-42be9ca0-468a-4b0f-8d63-ff311654795c.png)

Since the relative feature changes were small, it's a bit hard to test them specifically, however I am setting up a compose key to try using that under kmscon.

AndersonTorres reviewed (approved):
Waiting for the next meson.

omasanori reviewed (approved):
While I have still been building the system on my old laptop, I reviewed changes between two commits briefly and finds no huge and risky changes.

Feel free to merge this when meson is updated even if I have not yet reported the result on my machine at that moment. I will test as soon as possible though.

omasanori commented:
I had tested underline, italic and multilingual text with `emacs-nox`, `noto-fonts` and `noto-fonts-cjk-sans` and confirmed that it works. Thanks @Infinidoge !

Mindavi commented on pkgs/os-specific/linux/kmscon/default.nix:
```diff
@@ -1,46 +1,66 @@
+
+  # _FORTIFY_SOURCE requires compiling with optimization (-O)
+  NIX_CFLAGS_COMPILE = lib.optionalString stdenv.cc.isGNU "-O"
+    + " -Wno-error=maybe-uninitialized"; # https://github.com/Aetf/kmscon/issues/49
```
In my opinion, downstream packagers should never compile with Werror, since it can break on compiler updates. That's ubdesirable since a compiler update (generally) won't make the source code less safe in any way. And the compiler shouldn't be assumed anyway since different distros may use different compilers.

Mindavi replied:
So my suggestion is to add a blanket -Wno-error (or set a flag that does that)

Infinidoge commented:
Since the dependent PR has been merged, is there anything that needs to be done before this can be merged?

Mindavi commented:
@ofborg eval

Infinidoge commented:
Also, it seems that the meson changes have made their way into `master`, should I go ahead and retarget to there, since this change itself doesn't cause a mass rebuild?

Infinidoge commented:
Oh wait it already is targetting master, nevermind
